hardware_interface/src/resource_manager.cpp

```diff
@@ -616,7 +636,66 @@ bool ResourceManager::state_interface_is_available(const std::string & name) con
            name) != resource_storage_->available_state_interfaces_.end();
 }
 
-// CM API
+void ResourceManager::import_controller_reference_interfaces(
+  const std::string & controller_name, std::vector<CommandInterface> & interfaces)
+{
+  std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
+  std::lock_guard<std::recursive_mutex> guard_claimed(claimed_command_interfaces_lock_);
+  auto interface_names = resource_storage_->add_command_interfaces(interfaces);
+  resource_storage_->controllers_reference_interfaces_map_[controller_name] = interface_names;
+}
+
+std::vector<std::string> ResourceManager::get_controller_reference_interface_names(
+  const std::string & controller_name)
+{
+  return resource_storage_->controllers_reference_interfaces_map_.at(controller_name);
+}
+
+void ResourceManager::make_controller_reference_interfaces_available(
+  const std::string & controller_name)
+{
+  auto interface_names =
+    resource_storage_->controllers_reference_interfaces_map_.at(controller_name);
+  std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
+  resource_storage_->available_command_interfaces_.insert(
+    resource_storage_->available_command_interfaces_.end(), interface_names.begin(),
+    interface_names.end());
+}
+
+void ResourceManager::make_controller_reference_interfaces_unavailable(
+  const std::string & controller_name)
+{
+  auto interface_names =
+    resource_storage_->controllers_reference_interfaces_map_.at(controller_name);
+
+  std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
+  for (const auto & interface : interface_names)
+  {
+    auto found_it = std::find(
+      resource_storage_->available_command_interfaces_.begin(),
+      resource_storage_->available_command_interfaces_.end(), interface);
+    if (found_it != resource_storage_->available_command_interfaces_.end())
+    {
+      resource_storage_->available_command_interfaces_.erase(found_it);
+      RCUTILS_LOG_DEBUG_NAMED(
+        "resource_manager", "'%s' command interface removed from available list",
+        interface.c_str());
+    }
+  }
+}
+
+void ResourceManager::remove_controller_reference_interfaces(const std::string & controller_name)
+{
+  auto interface_names =
+    resource_storage_->controllers_reference_interfaces_map_.at(controller_name);
+  resource_storage_->controllers_reference_interfaces_map_.erase(controller_name);
+
+  std::lock_guard<std::recursive_mutex> guard(resource_interfaces_lock_);
```

Comment:
in other cases you put the guards into the function that is called. Is this different for any reason?

Reply:
Yes, because the methods called are not part of the `ResourceManager` class, but `ResourceStorage` and `Manager` is taking care of access to the storage.

Reply:
thx